kitodo-mediaserver-importer/src/main/java/org/kitodo/mediaserver/importer/control/ImporterFlowControl.java

```diff
@@ -55,111 +60,176 @@ public void setImporterProperties(ImporterProperties importerProperties) {
+
+                    Files.move(mets.toPath(), Paths.get(mets.getParent(), newWork.getId() + ".xml"));
+                    Path newPath = Paths.get(workDir.getParent(), newWork.getId());
+                    Files.move(workDir.toPath(), newPath);
+                    workDir = newPath.toFile();
+                }
+
+                // Check in the database if this work is already present
+                // and if there are identifiers associated to another work.
+                presentWork = workChecker.check(newWork);
+
+                // If the work is already present, it should be replaced.
+                if (presentWork != null) {
+
+                    LOGGER.info("Work " + newWork.getId() + " already present, replacing");
+
+                    newWork.setEnabled(presentWork.isEnabled());
+
+                    // Files created and cached by the fileserver must be deleted. TODO call action instead?
+                    fileDeleter.delete(Paths.get(fileserverProperties.getCachePath(), newWork.getId()));
+
+                    // Move old work files to a temporary folder.
+                    if (new File(presentWork.getPath()).isDirectory()) {
+                        tempOldWorkFiles = Paths.get(importerProperties.getTempWorkFolderPath(), presentWork.getId());
+                        Files.move(
+                                Paths.get(presentWork.getPath()),
+                                tempOldWorkFiles
+                        );
+                    } else {
+                        LOGGER.warn("The alleged root path " + presentWork.getPath() + " of the already present work "
+                                + presentWork.getId() + " is not a directory. Old files cannot be moved "
+                                + "to temporary folder");
+                    }
+                }
+
+                // Move work files to the production root.
+                Files.move(
+                        workDir.toPath(),
+                        Paths.get(newWork.getPath())
+                );
+
+                // TODO Perform all defined direct import actions (doi registration…).
+
+                // TODO Order all defined asynchronous import actions (creation of additional files…).
+
+
+                // Insert the work data into the database, updating if old data present.
+                workService.importWork(newWork);
+
+                // Delete temporary files.
+                if (tempOldWorkFiles != null) {
+                    fileDeleter.delete(tempOldWorkFiles);
+                }
+
+                // TODO Trigger indexing in presentation system (configurable).
+
+                LOGGER.info("Finished import of work " + workDir.getName());
+
+            } catch (Exception e) {
+                LOGGER.error("An error occured importing work " + workDir.getName()
+                        + ", moving all files to the error folder. Error: " + e);
+
+                // TODO this rollback has to be consolidated
```

Comment:
I'm not sure if a try{} over the whole process is a good way. For a rollback you need to know the current state or you need a snapshot. I think about DB migrations with the single migrationsscripts. You have to go back every single step. So we would need to "record" every step and on an error move back every step.

The current catch block will not reliably work. But it's also a question how far we want go.

Reply:
Sehe es hier ähnlich.